### ISSUE TITLE
Incorrect bootstrap file generation for new self-built compilers on AL AMI

### ISSUE THREAD

**Issue by tobodner** [awslabs/aws-lambda-cpp#79]
I am building my project inside a Docker container with an image derivative of lambci/lambda-base:build (Amazon Linux 2018.03 AMI). The compiler is a self-built GCC 9.2.0 (with libstdc++.so.6.0.27, which is included in the artefact independent of the NO_LIBC flag).

The AL AMI in AWS Lambda has only packages for GCC 4.8.5 and 7.2.0 (with libstdc++ *.19 and *.24, respectively: https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html). 

With the NO_LIBC flag, I get the following bootstrap file:
```
#!/bin/bash
set -euo pipefail
export AWS_EXECUTION_ENV=lambda-cpp
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$LAMBDA_TASK_ROOT/lib
exec $LAMBDA_TASK_ROOT/bin/myApp ${_HANDLER}
```
Together with the following error upon execution:
```
START RequestId: c63aad98-228d-4081-9475-2b7c884b00e6 Version: $LATEST
/var/task/bin/myApp: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /var/task/lib/libaws-cpp-sdk-core.so)
END RequestId: c63aad98-228d-4081-9475-2b7c884b00e6
REPORT RequestId: c63aad98-228d-4081-9475-2b7c884b00e6	Duration: 288.41 ms	Billed Duration: 300 ms	Memory Size: 128 MB	Max Memory Used: 15 MB	
RequestId: c63aad98-228d-4081-9475-2b7c884b00e6 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

Without the flag, I get:
```
#!/bin/bash
set -euo pipefail
export AWS_EXECUTION_ENV=lambda-cpp
exec $LAMBDA_TASK_ROOT/lib/ld-linux-x86-64.so.2 --library-path $LAMBDA_TASK_ROOT/lib $LAMBDA_TASK_ROOT/bin/myApp ${_HANDLER}
```
and a passing Lambda function.

Now, if I overwrite the bootstrap file in case #1 with the one produced for case #2 (while still only shipping the reduced set of libs from case #1), the Lambda function also passes.

Can we somehow enable this behaviour for compilers with newer standard C++ library implementations?

Thanks and best,
Thomas

**Comment by marcomagdy** [comment]
This is not a bug. This is by design.

The idea is that if you are building on Amazon Linux<sup>[1]</sup>, then you don't need to package glibc because that's the _same_ environment Lambda executes your code inside. Therefore glibc is guaranteed to work and also libstdc++ which is linked against that same glibc.

If I understand what you said correctly, you are building GCC 9.2.0 from source and using a prebuilt libstdc++. That will not work, because that libstdc++ is (evidently) linked against a newer version of glibc than the one shipping with AL.

So, in summary, `NO_LIBC` is meant to be used if you're building the runtime and the Lambda function on AL **without** using external prebuilt dependencies.

You can use prebuilt binaries only if you can guarantee they're linked (transitively) against the same or older version of glibc (glibc is pretty good about staying backwards compatible).

<sup>[1]</sup> Amazon Linux is different from Amazon Linux 2. I have to check if Lambda has moved to AL2 completely or not.

**Comment by tobodner** [comment]
Thanks for the explanation, Marco. I should have known that my version of libstdc++ is linked against a newer version of glibc.

I got two follow-up questions here:
- Why do you think I get away with the bootstrap file modified to use the dynamic linker? Am I linking here "at my own risk", betting that I only really need the safe subset of my newer glibc?
- How can one confidently check for the AL version in Lambda? I read https://aws.amazon.com/blogs/compute/upcoming-updates-to-the-aws-lambda-execution-environment/ plus checked under /etc/system-release in my recent Lambda executions.

Thanks and best,
Thomas

**Comment by marcomagdy** [comment]
>Why do you think I get away with the bootstrap file modified to use the dynamic linker

Most probably because you're now using the "right" dynamic loader (ld.so). But it's hard for me to answer definitively without seeing your zip package. I'd personally be uncomfortable shipping that.

>How can one confidently check for the AL version in Lambda?

The [documentation is not clear](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html) when it comes to custom runtimes (as such is the case with C++). It's definitely safer to assume they're running AL (as opposed to AL2). Maybe @bmoffatt has a better answer?


Lastly, I just would like to say that the whole point of this C++ runtime was to relieve users from dealing with such complicated details. You shouldn't have to worry about `glibc` or `libstdc++` versions and their interaction with Lambda's execution environment. My recommendation is to either use AL as is with NO_LIBC, or if you need a modern `libstdc++` implementation, use a different Linux distribution that ships with that `libstdc++` version and build your Lambda function there.

**Comment by tobodner** [comment]
Thanks again for your quick responses! This is much appreciated.

I am aware of the fact that I am - in part - working around the benefits of the C++ Lambda runtime. My motivation for this is my research in database systems on serverless infrastructure. I am trying to cut down on function package size to reduce cold start times of functions. My solution to above issue is to now go with a modern Clang version that links against the legacy libstdc++ version inside the Lambda execution environment.
